flink-table/flink-sql-gateway/src/test/java/org/apache/flink/table/gateway/service/context/SessionContextTest.java

```diff
@@ -133,6 +138,79 @@ public void testSetAndResetArbitraryKey() {
         assertThat(sessionContext.getConfigMap().get("bb")).isNull();
     }
 
+    @Test
+    public void testStatementSetStateTransition() {
+        assertThat(sessionContext.isStatementSetState()).isFalse();
+
+        Configuration emptyConfiguration = new Configuration();
+        TableEnvironmentInternal tableEnv =
+                sessionContext.createOperationExecutor(emptyConfiguration).getTableEnvironment();
+        tableEnv.executeSql("CREATE TABLE whatever (f STRING) WITH ('connector' = 'values')");
+
+        sessionContext
+                .createOperationExecutor(emptyConfiguration)
+                .executeStatement(OperationHandle.create(), "BEGIN STATEMENT SET;");
+
+        assertThat(sessionContext.isStatementSetState()).isTrue();
+
+        // invalid statement in Statement Set
+        assertThatThrownBy(
+                        () ->
+                                sessionContext
+                                        .createOperationExecutor(emptyConfiguration)
+                                        .executeStatement(OperationHandle.create(), "SELECT 1;"))
+                .satisfies(
+                        FlinkAssertions.anyCauseMatches(
+                                SqlExecutionException.class,
+                                "Wrong statement after 'BEGIN STATEMENT SET'.\n"
+                                        + "Only 'INSERT/CREATE TABLE AS' statement is allowed in Statement Set or use 'END' statement to terminate Statement Set."));
+
+        // 'isStatementSetState' is still true, and nothing in 'statementSetOperations'
+        assertThat(sessionContext.isStatementSetState()).isTrue();
+        assertThat(sessionContext.getStatementSetOperations().size()).isEqualTo(0);
+
+        // valid statement in Statement Set
+        sessionContext
+                .createOperationExecutor(emptyConfiguration)
+                .executeStatement(
+                        OperationHandle.create(), "INSERT INTO whatever VALUES('Hello World');");
```

Comment:
Is it worth testing multiple statements in the statement set?  Maybe the integration test covers that?

Reply:
Sure it is worth. I forgot to test this case. The test is improved now.